File: springboot-api/springboot-api/src/test/java/com/mehmetkaradana/springboot_api/loadtest/LoadTest.java
package com.mehmetkaradana.springboot_api.loadtest;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

public class LoadTest {

    private static final int TOTAL_REQUESTS = 1000; // Toplam istek sayısı
    private static final int THREAD_COUNT = 100; // Aynı anda gönderilecek istek sayısı
    private static final String URL = "http://localhost:8081/api/getRooms"; // Test etmek istediğiniz endpoint
    private static final RestTemplate restTemplate = new RestTemplate();

    public static void main(String[] args) {
        ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
        List<Future<Long>> responseTimes = new ArrayList<>();

        // Toplam istek sayısı kadar işlem oluştur
        for (int i = 0; i < TOTAL_REQUESTS; i++) {
            Future<Long> responseTime = executor.submit(() -> sendRequestAndMeasureTime());
            responseTimes.add(responseTime);
        }

        executor.shutdown();

        analyzeResults(responseTimes);
    }

    private static Long sendRequestAndMeasureTime() {
        try {
            long start = System.currentTimeMillis();

            HttpHeaders headers = new HttpHeaders();
            headers.set("Authorization", "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjY3MWEyOTFiNzY0YThlMTEwNzlkMGYyZSIsImlhdCI6MTcyOTc2NzcwNywiZXhwIjoxNzMyMzU5NzA3fQ.ijMcqGw4OUOWG79jrXhmkAwGolVJUjNXp6Y15y-OY7Y"); // Token'i buraya ekleyin
            HttpEntity<String> entity = new HttpEntity<>(headers);

            ResponseEntity<String> response = restTemplate.exchange(URL, HttpMethod.GET, entity, String.class);

            long end = System.currentTimeMillis();

            if (response.getStatusCode().is2xxSuccessful()) {
                return end - start;
            } else {
                System.out.println("Error: " + response.getStatusCode());
                return null;
            }

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
            return null;
        }
    }

    private static void analyzeResults(List<Future<Long>> responseTimes) {
        List<Long> times = new ArrayList<>();
        int errorCount = 0;

        for (Future<Long> future : responseTimes) {
            try {
                Long time = future.get();
                if (time != null) {
                    times.add(time);
                } else {
                    errorCount++;
                }
            } catch (Exception e) {
                errorCount++;
            }
        }

        long total = times.stream().mapToLong(Long::longValue).sum();
        long maxTime = times.stream().mapToLong(Long::longValue).max().orElse(0);
        long minTime = times.stream().mapToLong(Long::longValue).min().orElse(0);

        System.out.println("Total Requests: " + (times.size() + errorCount));
        System.out.println("Successful Requests: " + times.size());
        System.out.println("Failed Requests: " + errorCount);
        System.out.println("Average Response Time: " + (times.size() > 0 ? (total / times.size()) : 0) + " ms");
        System.out.println("Maximum Response Time: " + maxTime + " ms");
        System.out.println("Minimum Response Time: " + minTime + " ms");

    }
}

File: ktor-api/src/test/kotlin/com/mehmetkaradana/loadtest/LoadTest.kt
package com.mehmetkaradana.loadtest

import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicInteger
import kotlin.system.measureTimeMillis

object LoadTest {
    private const val TOTAL_REQUESTS = 1000 // Toplam istek sayısı
    private const val CONCURRENT_REQUESTS = 100 // Aynı anda gönderilecek istek sayısı
    private const val URL = "http://localhost:8082/api/getRooms" // Test etmek istediğiniz endpoint

    private val client = HttpClient()

    @JvmStatic
    fun main(args: Array<String>) = runBlocking {
        val responseTimes = mutableListOf<Long>()
        val errorCount = AtomicInteger(0)

        for (i in 0 until TOTAL_REQUESTS step CONCURRENT_REQUESTS) {
            val jobs = List(CONCURRENT_REQUESTS) {
                launch {
                    val responseTime = sendRequestAndMeasureTime()
                    if (responseTime != null) {
                        synchronized(responseTimes) { // Eşzamanlı erişimi kontrol et
                            responseTimes.add(responseTime)
                        }
                    } else {
                        errorCount.incrementAndGet()
                    }
                }
            }

            // Her grup için tüm işlerin tamamlanmasını bekle
            jobs.forEach { it.join() }
        }

        analyzeResults(responseTimes, errorCount.get())
    }

    private suspend fun sendRequestAndMeasureTime(): Long? {
        return try {
            measureTimeMillis {
                val response: HttpResponse = client.get(URL) {
                    header(HttpHeaders.Authorization, "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjY3MWEyOTFiNzY0YThlMTEwNzlkMGYyZSIsImlhdCI6MTcyOTc2NzcwNywiZXhwIjoxNzMyMzU5NzA3fQ.ijMcqGw4OUOWG79jrXhmkAwGolVJUjNXp6Y15y-OY7Y")
                }

                if (response.status.isSuccess()) {
                    // Yanıt başarılıysa, zamanın uzunluğunu döndür
                    return@measureTimeMillis
                } else {
                    println("Error: ${response.status}")
                    return null
                }
            }
        } catch (e: Exception) {
            println("Error: ${e.message}")
            null
        }
    }

    private fun analyzeResults(responseTimes: List<Long>, errorCount: Int) {
        val total = responseTimes.sum()
        val maxTime = responseTimes.maxOrNull() ?: 0
        val minTime = responseTimes.minOrNull() ?: 0

        println("Total Requests: ${responseTimes.size + errorCount}")
        println("Successful Requests: ${responseTimes.size}")
        println("Failed Requests: $errorCount")
        println("Average Response Time: ${if (responseTimes.isNotEmpty()) total / responseTimes.size else 0} ms") // Average response time
        println("Maximum Response Time: $maxTime ms")
        println("Minimum Response Time: $minTime ms")

    }
}
